### PR TITLE
347 Change world object to set migration rules vector only one time

### LINKED ISSUE #347: Set migration rules vector only one time.
In the function World::migration in world.cpp we set the migration rules each time this function get's called (see below).  This shouldn't be necessary as we can set them one time.

`  if (m_use_migration_rules) {
        rules = {
            std::make_pair(&return_home_when_recovered,
                           std::vector<LocationType>{
                               LocationType::Home,
                               LocationType::Hospital}), //assumption: if there is an ICU, there is also an hospital
            std::make_pair(&go_to_hospital, std::vector<LocationType>{LocationType::Home, LocationType::Hospital}),
            std::make_pair(&go_to_icu, std::vector<LocationType>{LocationType::Hospital, LocationType::ICU}),
            std::make_pair(&go_to_school, std::vector<LocationType>{LocationType::School, LocationType::Home}),
            std::make_pair(&go_to_work, std::vector<LocationType>{LocationType::Home, LocationType::Work}),
            std::make_pair(&go_to_shop, std::vector<LocationType>{LocationType::Home, LocationType::BasicsShop}),
            std::make_pair(&go_to_event, std::vector<LocationType>{LocationType::Home, LocationType::SocialEvent}),
            std::make_pair(&go_to_quarantine, std::vector<LocationType>{LocationType::Home})};
    }
    else {
        rules = {
            std::make_pair(&return_home_when_recovered,
                           std::vector<LocationType>{
                               LocationType::Home,
                               LocationType::Hospital}), //assumption: if there is an ICU, there is also an hospital
            std::make_pair(&go_to_hospital, std::vector<LocationType>{LocationType::Home, LocationType::Hospital}),
            std::make_pair(&go_to_icu, std::vector<LocationType>{LocationType::Hospital, LocationType::ICU}),
            std::make_pair(&go_to_quarantine, std::vector<LocationType>{LocationType::Home})};
    }`

### PR DESCRIPTION
**Pull request by khoanguyen-dev**
## Merge Request - GuideLine Checklist 

**Guideline** to check code before resolve WIP and approval, respectively.
As many checkboxes as possible should be ticked.

### Checks by code author:
* [x] There is at least one issue associated with the pull request.
* [x] The branch follows the naming conventions as defined in the [git workflow](git-workflow).
* [x] New code adheres with the [coding guidelines](coding-guidelines)
* [x] Tests for new functionality has been added
* [x] A local test was succesful
* [x] There is appropriate **documentation** of your work. (use doxygen style comments)
* [x] If new third party software is used, did you pay attention to its license? Please remember to add it to the wiki after successful merging.
* [x] If new mathematical methods or epidemiological terms are used, has the glossary been updated ? Did you provide further documentation ?
 is present or referenced. Please provide your references.
* [x] The following questions are addressed in the documentation*:  Developers (what did you do?, how can it be maintained?), For users (how to use your work?), For admins (how to install and configure your work?)
* For documentation: Please write or update the Readme in the current working directory!

### Checks by code reviewer(s):
* [x] Is the code clean of development artifacts e.g., unnecessary comments, prints, ...
* [ ] The ticket goals for each associated issue are reached or problems are clearly addressed (i.e., a new issue was introduced).
* [ ] There are appropriate **unit tests** and they pass.
* [ ] The git history is clean and linearized for the merge request.
* [x] Coverage report for new code is acceptable. 

closes #347

### CODE REVIEW

File: cpp/models/abm/world.h [hunk 1]
```diff
@@ -195,6 +195,8 @@ class World
     MigrationParameters m_migration_parameters;
     TripList m_trip_list;
     bool m_use_migration_rules;
+    using migration_rule = LocationType (*)(const Person&, TimePoint, TimeSpan, const MigrationParameters&);
+    std::vector<std::pair<migration_rule, std::vector<LocationType>>> global_migration_rules;
```
Comment (mknaranja) [inline cpp/models/abm/world.h:199]: please use m_* for member variables; see also https://github.com/DLR-SC/memilio/wiki/cpp-coding-guidelines

File: cpp/models/abm/world.h [hunk 2]
```diff
@@ -59,8 +59,8 @@ class World
         , m_infection_parameters(params)
         , m_migration_parameters()
         , m_trip_list()
-        , m_use_migration_rules(true)
     {
+        use_migration_rules(true);
```
Comment (mknaranja) [inline cpp/models/abm/world.h:63]: We should maybe make a new issue (if it does not already exist) to implement the possibility to not use these rules. (This shouldnt be done in this issue)

**Review by mknaranja — COMMENTED** [review]
Thanks @khoanguyen-dev. Obviously you made yourself familiar with the code quite fast :)

I would also like @xsaschako or @DavidKerkmann to also have a look but I think the goal of the issue is attained and it can probably be merged from my side (if you change this minor item on the member variable).

File: cpp/models/abm/world.h (hunk 1 above)
Comment (khoanguyen-dev) [reply cpp/models/abm/world.h]: Thanks for your comment. It's done and pushed.

**Review by DavidKerkmann — APPROVED** [review]
Looks good to me. Please always add the number of the issue also in the name of the pull request, i. e., `347 Change world object to set migration rules vector only one time`.

File: cpp/models/abm/world.h (hunk 2 above)
Comment (mknaranja) [reply cpp/models/abm/world.h:63]: see #447

**Comment by mknaranja** [comment]
Congratulations @khoanguyen-dev, first pull request merged ;)

**Pull request merged**